File: Leetcode/Leetcode-Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> sol = new ArrayList<Integer>();
        TreeNode cur = root;
        TreeNode pre = root;
        while(cur != null){
            if(cur.left == null){
                sol.add(cur.val);
                cur = cur.right;
            }
            else{
                pre = cur.left;
                //get the rightmost child
                while(pre.right != null){
                    pre = pre.right;
                }
                pre.right = cur;
                TreeNode temp = cur;
                cur = cur.left;
                temp.left = null;
            }
        }
        return sol;
    }
}

File: Leetcode/Leetcode - Binary Tree Inorder Traversal.py
"""
Given a binary tree, return the inorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
    \
     2
    /
   3

Output: [1,3,2]
Follow up: Recursive solution is trivial, could you do it iteratively?
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        ans = []
        stack = [root]
        while stack:
            cur = stack[-1]
            if not cur.left: ans.append(stack.pop().val)
            if cur.left:
                stack.append(cur.left)
                cur.left = None
                continue
            elif cur.right:
                stack.append(cur.right)
        return ans
